synapseclient/core/download/download_functions.py

```diff
@@ -608,6 +646,24 @@ def download_fn(
                     ),
                 )
 
+                # Record span event for non-multipart download completion
+                file_size = (
+                    os.path.getsize(downloaded_path)
+                    if os.path.exists(downloaded_path)
+                    else 0
+                )
+                span.add_event(
+                    "download_chunk_completed",
+                    {
+                        "chunk_number": 0,
+                        "start_byte": 0,
+                        "end_byte": file_size - 1 if file_size > 0 else 0,
+                        "file_handle_id": file_handle_id,
+                        "synapse_id": synapse_id,
+                        "time_to_transfer_seconds": time.time() - download_start_time,
+                    },
+                )
```

Comment:
It's too bad these specific chunks of code are repeated, I wonder if a way-down-the-line optimization is to have a factory class for downloading, specifically around supporting the different cloud storage buckets.

That said - that may be overkill or the wrong design choice. Thoughts?

Reply:
I like that idea of refactoring it eventually to follow that kind of pattern. I dislike how cognitively complex this operation is.